### ISSUE TITLE
 In-memory engine: assertion failed during prewrite

### ISSUE THREAD

**Issue by overvenus** [tikv/tikv#17074]
## Bug Report

TPCC log

```
[2024-05-22 18:56:33] execute run failed, err Error 8141: assertion failed: key: 7480000000000000725f72038000000000000001038000000000000008038000000000004e4f, assertion: Exist, start_ts: 449945136815407106, existing start ts: 0, existing commit ts: 0
execute run failed, err Error 8141: assertion failed: key: 7480000000000000725f72038000000000000001038000000000000008038000000000004e4f, assertion: Exist, start_ts: 449945136815407106, existing start ts: 0, existing commit ts: 0
```

TiDB log

```log
[2024/05/22 18:56:33.212 +00:00] [ERROR] [session.go:653] ["assertion failed"] [conn=3986714032] [session_alias=] [message="[tikv:8141]assertion failed: key: 7480000000000000725f72038000000000000001038000000000000008038000000000004e4f, assertion: Exist, start_ts: 449945060754587683, existing start ts: 0, existing commit ts: 0"] ["mvcc history"="{\"decoded\":{\"449938179959292054\":{}},\"key\":\"7480000000000000725F72038000000000000001038000000000000008038000000000004E4F\",\"mvcc\":{\"info\":{\"lock\":{\"type\":1,\"start_ts\":449945060754587683,\"primary\":\"dIAAAAAAAAByX3IDgAAAAAAAAAEDgAAAAAAAAAEDgAAAAAAAaY0=\"},\"writes\":[{\"type\":1,\"start_ts\":449939751971324000,\"commit_ts\":449939752036860048},{\"start_ts\":449938179959292054,\"commit_ts\":449938179972399159,\"short_value\":\"gAAAAAAA\"}]}},\"regionID\":5838}"]
```

MVCC history

```json
{
    "decoded": {
        "449938179959292054": {}
    },
    "key": "7480000000000000725F72038000000000000001038000000000000008038000000000004E4F",
    "mvcc": {
        "info": {
            "lock": {
                "type": 1, // LockType::Delete
                "start_ts": 449945060754587683, // 2024-05-22 18:56:32.942 +0000 UTC
                "primary": "dIAAAAAAAAByX3IDgAAAAAAAAAEDgAAAAAAAAAEDgAAAAAAAaY0="
            },
            "writes": [
                {
                    "type": 1, // WriteType::Delete
                    "start_ts": 449939751971324000, // 2024-05-22 13:19:01.541 +0000 UTC
                    "commit_ts": 449939752036860048 // 2024-05-22 13:19:01.791 +0000 UTC
                },
                {   // type: 0 // WriteType::Put
                    "start_ts": 449938179959292054, // 2024-05-22 11:39:04.791 +0000 UTC
                    "commit_ts": 449938179972399159, // 2024-05-22 11:39:04.841 +0000 UTC
                    "short_value": "gAAAAAAA" // [128 0 0 0 0 0] See https://go.dev/play/p/CU2nj4mI9be
                }
            ]
        }
    },
    "regionID": 5838
}
```

TiKV log

```
[2024/05/22 18:53:14.596 +00:00] [INFO] [apply.rs:1714] ["execute admin command"] [command="cmd_type: BatchSplit splits { requests { split_key: 7480000000000000FF725F720380000000FF0000000203800000FF0000000002000000FC new_region_id: 5838 new_peer_ids: 5839 } right_derive: true }"] [index=434861] [term=6] [peer_id=4269] [region_id=4268] [thread_id=145]
[2024/05/22 18:53:14.596 +00:00] [INFO] [apply.rs:2586] ["split region"] [keys="key 7480000000000000FF725F720380000000FF0000000203800000FF0000000002000000FC"] [region="id: 4268 start_key: 7480000000000000FF7200000000000000F8 end_key: 7480000000000000FF725F720380000000FF0000000303800000FF0000000005038000FF000000003F7D0000FD region_epoch { conf_ver: 1 version: 72 } peers { id: 4269 store_id: 1 }"] [peer_id=4269] [region_id=4268] [thread_id=145]
[2024/05/22 18:53:14.599 +00:00] [INFO] [peer.rs:4385] ["insert new region"] [store_id=1] [is_uninitialized_peer_exist=false] [region="id: 5838 start_key: 7480000000000000FF7200000000000000F8 end_k>[2024/05/22 18:53:14.599 +00:00] [INFO] [peer.rs:276] ["create peer"] [peer_id=5839] [region_id=5838] [thread_id=144]
[2024/05/22 18:53:14.599 +00:00] [INFO] [raft.rs:2684] ["switched to configuration"] [config="Configuration { voters: Configuration { incoming: Configuration { voters: {5839} }, outgoing: Configura>[2024/05/22 18:53:14.599 +00:00] [INFO] [raft.rs:1151] ["became follower at term 5"] [term=5] [raft_id=5839] [region_id=5838] [thread_id=144]
[2024/05/22 18:53:14.599 +00:00] [INFO] [raft.rs:387] [newRaft] [peers="Configuration { incoming: Configuration { voters: {5839} }, outgoing: Configuration { voters: {} } }"] ["last term"=5] ["last>[2024/05/22 18:53:14.600 +00:00] [INFO] [raw_node.rs:315] ["RawNode created with id 5839."] [id=5839] [raft_id=5839] [region_id=5838] [thread_id=144]
[2024/05/22 18:53:14.600 +00:00] [INFO] [raft.rs:1556] ["starting a new election"] [term=5] [raft_id=5839] [region_id=5838] [thread_id=144]
[2024/05/22 18:53:14.600 +00:00] [INFO] [raft.rs:1201] ["became pre-candidate at term 5"] [term=5] [raft_id=5839] [region_id=5838] [thread_id=144]
[2024/05/22 18:53:14.600 +00:00] [INFO] [raft.rs:1175] ["became candidate at term 6"] [term=6] [raft_id=5839] [region_id=5838] [thread_id=144]
[2024/05/22 18:53:14.600 +00:00] [INFO] [raft.rs:1259] ["became leader at term 6"] [term=6] [raft_id=5839] [region_id=5838] [thread_id=144]
[2024/05/22 18:53:14.600 +00:00] [INFO] [peer.rs:5712] ["require updating max ts"] [initial_status=25769803922] [region_id=5838] [thread_id=144]
[2024/05/22 18:53:14.600 +00:00] [INFO] [pd.rs:1705] ["succeed to update max timestamp"] [region_id=5838] [thread_id=37]
[2024/05/22 18:53:14.600 +00:00] [INFO] [endpoint.rs:358] ["Resolver initialized"] [pending_data_index=0] [snapshot_index=434861] [observe_id=ObserveId(5142)] [region=4268] [thread_id=106]
[2024/05/22 18:53:14.605 +00:00] [INFO] [endpoint.rs:701] ["register observe region"] [region="id: 5838 start_key: 7480000000000000FF7200000000000000F8 end_key: 7480000000000000FF725F720380000000FF0000000203800000FF0000000002000000FC region_epoch { conf_ver: 1 version: 73 } peers { id: 5839 store_id: 1 }"] [thread_id=106]
...
[2024/05/22 18:56:31.684 +00:00] [INFO] [background.rs:792] [load_evict] [may_evict="[]"] [ranges_to_add="[CacheRange { tag: \"[region_id=2]\", range_start: 7A7800000100000000FB, range_end: 7B }, CacheRange { tag: \"[region_id=5820]\", range_start: 7A7480000000000000FF765F720380000000FF0000000603800000FF0000000002038000FF0000000007C40380FF0000000000000A00FE, range_end: 7A7480000000000000FF765F720380000000FF0000000603800000FF0000000003038000FF000000002B9B0380FF0000000000000900FE }, CacheRange { tag: \"[region_id=5822]\", range_start: 7A7480000000000000FF785F720380000000FF0000004F03800000FF0000015AF1000000FC, range_end: 7A7480000000000000FF785F720380000000FF0000004F03800000FF0000017A91000000FC }, CacheRange { tag: \"[region_id=5824]\", range_start: 7A7480000000000000FF785F720380000000FF0000003F03800000FF0000015E2C000000FC, range_end: 7A7480000000000000FF785F720380000000FF0000003F03800000FF0000017D63000000FC }, CacheRange { tag: \"[region_id=5826]\", range_start: 7A7480000000000000FF765F720380000000FF0000003B03800000FF0000000008038000FF0000000008EB0380FF0000000000000100FE, range_end: 7A7480000000000000FF765F720380000000FF0000003B03800000FF0000000009038000FF000000002BDD0380FF0000000000000100FE }, CacheRange { tag: \"[region_id=5828]\", range_start: 7A7480000000000000FF765F720380000000FF0000004B03800000FF0000000006038000FF0000000002340380FF0000000000000400FE, range_end: 7A7480000000000000FF765F720380000000FF0000004B03800000FF0000000007038000FF0000000028EE0380FF0000000000000200FE }, CacheRange { tag: \"[region_id=5830]\", range_start: 7A7480000000000000FF765F720380000000FF0000005703800000FF0000000006038000FF00000000015E0380FF0000000000000200FE, range_end: 7A7480000000000000FF765F720380000000FF0000005703800000FF0000000007038000FF0000000028D10380FF0000000000000300FE }, CacheRange { tag: \"[region_id=5832]\", range_start: 7A7480000000000000FF765F720380000000FF0000002A03800000FF0000000002038000FF000000000C820380FF0000000000000B00FE, range_end: 7A7480000000000000FF765F720380000000FF0000002A03800000FF0000000003038000FF000000002FD40380FF0000000000000C00FE }, CacheRange { tag: \"[region_id=5834]\", range_start: 7A7480000000000000FF765F720380000000FF0000001203800000FF0000000007038000FF000000000BA80380FF0000000000000400FE, range_end: 7A7480000000000000FF765F720380000000FF0000001203800000FF0000000008038000FF000000002E390380FF0000000000000600FE }, CacheRange { tag: \"[region_id=5836]\", range_start: 7A7480000000000000FF785F720380000000FF0000005803800000FF000000FCD7000000FC, range_end: 7A7480000000000000FF785F720380000000FF0000005803800000FF0000011C10000000FC }, CacheRange { tag: \"[region_id=5838]\", range_start: 7A7480000000000000FF7200000000000000F8, range_end: 7A7480000000000000FF725F720380000000FF0000000203800000FF0000000002000000FC }, CacheRange { tag: \"[region_id=5840]\", range_start: 7A7480000000000000FF785F720380000000FF0000003903800000FF000000A3FB000000FC, range_end: 7A7480000000000000FF785F720380000000FF0000003903800000FF000000C342000000FC }, CacheRange { tag: \"[region_id=5842]\", range_start: 7A7480000000000000FF785F720380000000FF0000003403800000FF000000137D000000FC, range_end: 7A7480000000000000FF785F720380000000FF0000003403800000FF0000003163000000FC }, CacheRange { tag: \"[region_id=5844]\", range_start: 7A7480000000000000FF785F720380000000FF0000001803800000FF0000016143000000FC, range_end: 7A7480000000000000FF785F720380000000FF0000001803800000FF0000018117000000FC }, CacheRange { tag: \"[region_id=5846]\", range_start: 7A7480000000000000FF765F720380000000FF0000004E03800000FF0000000007038000FF0000000010670380FF0000000000000C00FE, range_end: 7A7480000000000000FF765F720380000000FF0000004E03800000FF0000000008038000FF00000000335A0380FF0000000000000D00FE }, CacheRange { tag: \"[region_id=5848]\", range_start: 7A7480000000000000FF765F720380000000FF0000001E03800000FF000000000A038000FF000000000BA80380FF0000000000000400FE, range_end: 7A7480000000000000FF765F720380000000FF0000001F03800000FF0000000001038000FF000000002EBF0380FF0000000000000300FE }, CacheRange { tag: \"[region_id=5850]\", range_start: 7A7480000000000000FF765F720380000000FF0000006003800000FF0000000004038000FF0000000004720380FF0000000000000700FE, range_end: 7A7480000000000000FF765F720380000000FF0000006003800000FF0000000005038000FF0000000029820380FF0000000000000500FE }, CacheRange { tag: \"[region_id=5852]\", range_start: 7A7480000000000000FF785F720380000000FF0000001903800000FF0000005FBD000000FC, range_end: 7A7480000000000000FF785F720380000000FF0000001903800000FF0000007F40000000FC }, CacheRange { tag: \"[region_id=5854]\", range_start: 7A7480000000000000FF785F720380000000FF0000005003800000FF0000003FBE000000FC, range_end: 7A7480000000000000FF785F720380000000FF0000005003800000FF0000005F3F000000FC }]"] [thread_id=27]
...
[2024/05/22 18:56:31.696 +00:00] [INFO] [engine.rs:393] ["Range to load"] [Pending=1] [Cached=15] [Tag="[region_id=5838]"] [thread_id=146]
[2024/05/22 18:56:31.697 +00:00] [INFO] [background.rs:935] ["Loading range"] [range="CacheRange { tag: \"[region_id=5838]\", range_start: 7A7480000000000000FF7200000000000000F8, range_end: 7A7480000000000000FF725F720380000000FF0000000203800000FF0000000002000000FC }"] [thread_id=28]
...
[2024/05/22 18:56:32.607 +00:00] [INFO] [background.rs:1020] ["Loading range finished"] [duration(sec)=909.870954ms] [range="CacheRange { tag: \"[region_id=5838]\", range_start: 7A7480000000000000FF7200000000000000F8, range_end: 7A7480000000000000FF725F720380000000FF0000000203800000FF0000000002000000FC }"] [thread_id=28]
...
[2024/05/23 18:53:33.753 +00:00] [INFO] [background.rs:577] ["safe point update"] [range="CacheRange { tag: \"[region_id=5838]\", range_start: 7A7480000000000000FF7200000000000000F8, range_end: 7A7480000000000000FF725F720380000000FF0000000203800000FF0000000002000000FC }"] [current=449967615301910528] [prev=449967568115990528] [thread_id=31]
[2024/05/23 18:53:34.264 +00:00] [INFO] [background.rs:609] ["range gc complete"] [current_safe_point=449967615301910528] [below_safe_point_delete_version=0] [below_safe_point_version=2807902] [below_safe_point_unique_keys=2807902] [filtered_version=0] [total_version=2807902] [gc_duration=510.202059ms] [range="CacheRange { tag: \"[region_id=5838]\", range_start: 7A7480000000000000FF7200000000000000F8, range_end: 7A7480000000000000FF725F720380000000FF0000000203800000FF0000000002000000FC }"] [thread_id=31]
...
[2024/05/23 18:56:33.710 +00:00] [INFO] [background.rs:577] ["safe point update"] [range="CacheRange { tag: \"[region_id=5838]\", range_start: 7A7480000000000000FF7200000000000000F8, range_end: 7A7480000000000000FF725F720380000000FF0000000203800000FF0000000002000000FC }"] [current=449967662487830528] [prev=449967615301910528] [thread_id=31]
[2024/05/23 18:56:34.206 +00:00] [INFO] [background.rs:609] ["range gc complete"] [current_safe_point=449967662487830528] [below_safe_point_delete_version=0] [below_safe_point_version=2807902] [below_safe_point_unique_keys=2807902] [filtered_version=0] [total_version=2807902] [gc_duration=495.643897ms] [range="CacheRange { tag: \"[region_id=5838]\", range_start: 7A7480000000000000FF7200000000000000F8, range_end: 7A7480000000000000FF725F720380000000FF0000000203800000FF0000000002000000FC }"] [thread_id=31]
```

### What version of TiKV are you using?

Nightly, built on #16868

### Steps to reproduce

TiKV Config

```toml
[range-cache-engine]
  enabled = true
  hard-limit-threshold = "8GB"
  soft-limit-threshold = "5GB"
```

TPCC commands

```bash
/root/.tiup/components/bench/v1.12.0/go-tpc \
    tpcc prepare \
    -H $HOST -P $PORT \
    -D tpcc100 --warehouses 100 -T 50

# Run
/root/.tiup/components/bench/v1.12.0/go-tpc \
    tpcc run \
    -H $HOST -P $PORT \
    -D tpcc100 --warehouses 100 -T 50
```

### What did you expect?

No such error.

**Comment by cfzjywxk** [comment]
In correctness testing(or non performance benchmark testing), setting the assertion level(https://docs.pingcap.com/tidb/stable/system-variables#tidb_txn_assertion_level-new-in-v600) to `strict` can achieve more rigorous checks and thus discover issues more quickly.
/cc @SpadeA-Tang

**Comment by overvenus** [comment]
In the past month, we've run extensive TPCC tests, and the bug has not reoccurred. It’s likely been fixed by #17261 and #17288.
